light_curves/code_src/hats_functions.py

```diff
@@ -0,0 +1,133 @@
+    filter_id_to_name: dict[int, str],
+    id_col: str,
+    time_col: str,
+    flux_col: str,
+    err_col: str,
+    *,
+    radius: float = 1.0,
+    object_margin_cache=None,
+    lightcurve_margin_cache=None,
+) -> MultiIndexDFObject:
+    """
+    Generic LSDB/HATs light-curve fetcher by cross-matching an object catalog
+    to an input coord list, then joining to a light-curve catalog.
+
+    Parameters
+    ----------
+    sample_table : astropy.table.Table
+        Table with columns ['objectid', 'coord', 'label'] (SkyCoord in 'coord').
+    object_catalog_path : str
+        S3/HATs path to the **object** catalog (used for cross-matching).
+    lightcurve_catalog_path : str
+        S3/HATs path to the **light curve** catalog.
+    object_columns : list of str
+        Columns to read from the object catalog (must include `id_col`, ra/dec).
+    detect_columns : list of str
+        Columns to read from the light-curve catalog (must include `id_col`,
+        `time_col`, `flux_col`, `err_col`, plus the filter-ID column).
+    filter_id_to_name : dict
+        Mapping from integer filter IDs to human-readable band names.
+    id_col : str
+        Name of the object-ID column in both catalogs.
+    time_col : str
+        Name of the time/MJD column in the light-curve catalog.
+    flux_col : str
+        Name of the flux column in the light-curve catalog.
+    err_col : str
+        Name of the flux-uncertainty column in the light-curve catalog.
+    radius : float, optional
+        Cross-match radius in arcseconds (default 1.0).
+
+    Returns
+    -------
+    MultiIndexDFObject
+        A multi-indexed (objectid, label, band, time) container of the light curves.
+    """
+    # 1) read the two catalogs lazily
+    
+    # this table will be used for cross matching with our sample's ra and decs
+    # but does not have light curve information
+    obj_cat = lsdb.read_hats(object_catalog_path,
+                             columns=object_columns,
+                             margin_cache=object_margin_cache)
+    # this table houses the light curves
+    lc_cat  = lsdb.read_hats(lightcurve_catalog_path,
+                             columns=light_curve_columns,
+                            margin_cache=lightcurve_margin_cache)
+    
+    # 2) convert astropy sample_table into an LSDB catalog
+    sample_df = pd.DataFrame({
+        'objectid': sample_table['objectid'],
+        'ra_deg':    sample_table['coord'].ra.deg,
+        'dec_deg':   sample_table['coord'].dec.deg,
+        'label':     sample_table['label']
+    })
+    # convert dataframe to hats catalog
+    sample_lsdb = lsdb.from_dataframe(
+        sample_df,
+        ra_column='ra_deg',
+        dec_column='dec_deg',
+        margin_threshold=10,
+        drop_empty_siblings=True
+    )
+
+    # 3) cross-match to find nearest object
+    # only keep the best match
+    matched_obj = obj_cat.crossmatch(
+        sample_lsdb,
+        radius_arcsec=radius,
```

Comment:
```suggestion
    matched_obj = sample_lsdb.crossmatch(
        obj_cat,
        radius_arcsec=radius,
```

I think these catalogs should be reversed. Otherwise this will return more than one row per target if there is more than one catalog object within `radius` of it. I think what we actually want is the single catalog object that is closest to each target.

Reply:
IThis PR isn't going to go through, but just to add that I tried swapping the order for the ztf functions and I got an error.  I do have "n_neighbors = 1" in the next line which is what tells it to only return the best match.  Full traceback if you want to go through it is below.  I will not bother with figuring this out for the new PR.

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[4], line 5
      3 # get ZTF lightcurves
      4 ztf_search_radius = 1.0 #  arcsec
----> 5 df_lc_ZTF = ztf_get_lightcurves(sample_table, radius = ztf_search_radius)
      7 # add the resulting dataframe to all other archives
      8 df_lc.append(df_lc_ZTF)

File [~/fornax-demo-notebooks/light_curves/code_src/ztf_functions.py:86](https://science-console.fornax.smce.nasa.gov/jupyter/user/jkrick/lab/workspaces/fornax-demo-notebooks/light_curves/code_src/ztf_functions.py#line=85), in ztf_get_lightcurves(sample_table, radius)
     73 matched_lc = ztf_lc.crossmatch(
     74     ztf_lc,
     75     radius_arcsec=radius,
     76     n_neighbors=1,
     77 )
     78 #matched_lc = sample_lsdb.crossmatch(
     79 #    sample_lsdb,
     80 #    radius_arcsec=radius,
   (...)     84 # here is where the actual work gets done
     85 # compute the cross match with object table
---> 86 df = matched_lc.compute()
     88 #done with dask
     89 client.close() 

File [/opt/conda/envs/notebook/lib/python3.12/site-packages/lsdb/catalog/dataset/dataset.py:47](https://science-console.fornax.smce.nasa.gov/opt/conda/envs/notebook/lib/python3.12/site-packages/lsdb/catalog/dataset/dataset.py#line=46), in Dataset.compute(self)
     45 def compute(self) -> npd.NestedFrame:
     46     """Compute dask distributed dataframe to pandas dataframe"""
---> 47     return self._ddf.compute()

File [/opt/conda/envs/notebook/lib/python3.12/site-packages/lsdb/nested/core.py:437](https://science-console.fornax.smce.nasa.gov/opt/conda/envs/notebook/lib/python3.12/site-packages/lsdb/nested/core.py#line=436), in NestedFrame.compute(self, **kwargs)
    435 def compute(self, **kwargs):
    436     """Compute this Dask collection, returning the underlying dataframe or series."""
--> 437     return npd.NestedFrame(super().compute(**kwargs))

File [/opt/conda/envs/notebook/lib/python3.12/site-packages/dask/base.py:373](https://science-console.fornax.smce.nasa.gov/opt/conda/envs/notebook/lib/python3.12/site-packages/dask/base.py#line=372), in DaskMethodsMixin.compute(self, **kwargs)
    349 def compute(self, **kwargs):
    350     """Compute this dask collection
    351 
    352     This turns a lazy Dask collection into its in-memory equivalent.
   (...)    371     dask.compute
    372     """
--> 373     (result,) = compute(self, traverse=False, **kwargs)
    374     return result

File [/opt/conda/envs/notebook/lib/python3.12/site-packages/dask/base.py:681](https://science-console.fornax.smce.nasa.gov/opt/conda/envs/notebook/lib/python3.12/site-packages/dask/base.py#line=680), in compute(traverse, optimize_graph, scheduler, get, *args, **kwargs)
    678     expr = expr.optimize()
    679     keys = list(flatten(expr.__dask_keys__()))
--> 681     results = schedule(expr, keys, **kwargs)
    683 return repack(results)

File [/opt/conda/envs/notebook/lib/python3.12/site-packages/dask/dataframe/utils.py:357](https://science-console.fornax.smce.nasa.gov/opt/conda/envs/notebook/lib/python3.12/site-packages/dask/dataframe/utils.py#line=356), in check_meta()
    352     errmsg = "Expected partition of type `{}` but got `{}`".format(
    353         typename(type(meta)),
    354         typename(type(x)),
    355     )
    356 elif is_dataframe_like(meta):
--> 357     dtypes = pd.concat([x.dtypes, meta.dtypes], axis=1, sort=True)
    358     bad_dtypes = [
    359         (repr(col), a, b)
    360         for col, a, b in dtypes.fillna("-").itertuples()
    361         if not equal_dtypes(a, b)
    362     ]
    363     if bad_dtypes:

File [/opt/conda/envs/notebook/lib/python3.12/site-packages/pandas/core/indexes/base.py:4429](https://science-console.fornax.smce.nasa.gov/opt/conda/envs/notebook/lib/python3.12/site-packages/pandas/core/indexes/base.py#line=4428), in reindex()
   4426     raise ValueError("cannot handle a non-unique multi-index!")
   4427 elif not self.is_unique:
   4428     # GH#42568
-> 4429     raise ValueError("cannot reindex on an axis with duplicate labels")
   4430 else:
   4431     indexer, _ = self.get_indexer_non_unique(target)

ValueError: cannot reindex on an axis with duplicate labels

Reply:
I believe the error is happening because of the line below that passes `suffixes=("", "")` which is not advised because it cause this kind of confusion. (In this case, the library is confused. In other cases humans will be confused.)

Reply:
> I do have "n_neighbors = 1" in the next line which is what tells it to only return the best match.

But it will return the best match for every row in the catalog on the left, which means that rows in the catalog on the right might appear more than once in the results.